src/parse-nm.c

```diff
@@ -64,17 +65,306 @@ ap_type_from_str(const char* type_str)
+        /* Append gateway/via IP */
+        if (split[0] && split[1])
+            route->via = g_strdup(split[1]); //no need to free, will stay in netdef
+        /* Append metric */
+        if (split[0] && split[1] && split[2] && strtoul(split[2], NULL, 10) != NETPLAN_METRIC_UNSPEC)
+            route->metric = strtoul(split[2], NULL, 10);
+        g_strfreev(split);
+
+        /* Parse route options */
+        if (options_kf_value) {
+            g_debug("%s: adding new route_options (kf)", options_key);
+            split = g_strsplit(options_kf_value, ",", -1);
+            for (unsigned i = 0; split[i]; ++i) {
+                g_debug("processing route_option: %s", split[i]);
+                gchar **kv = g_strsplit(split[i], "=", 2);
+                if (g_strcmp0(kv[0], "onlink") == 0)
+                    route->onlink = (g_strcmp0(kv[1], "true") == 0);
+                else if (g_strcmp0(kv[0], "initrwnd") == 0)
+                    route->advertised_receive_window = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "initcwnd") == 0)
+                    route->congestion_window = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "mtu") == 0)
+                    route->mtubytes = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "table") == 0)
+                    route->table = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "src") == 0)
+                    route->from = g_strdup(kv[1]); //no need to free, will stay in netdef
+                else
+                    unhandled_data = TRUE;
+                g_strfreev(kv);
+            }
+            g_strfreev(split);
+
+            if (!unhandled_data)
+                _kf_clear_key(kf, group, options_key);
+            g_free(options_key);
+            g_free(options_kf_value);
+        }
+
+        /* Add route to array, clear keyfile */
+        g_array_append_val(*routes_arr, route);
+        if (!unhandled_data)
+            _kf_clear_key(kf, group, key);
+        g_free(key);
+        g_free(kf_value);
+    }
+}
+
+static void
+parse_dhcp_overrides(GKeyFile* kf, const gchar* group, NetplanDHCPOverrides* dataptr)
+{
+    if (   g_key_file_get_boolean(kf, group, "ignore-auto-routes", NULL)
+        && g_key_file_get_boolean(kf, group, "never-default", NULL)) {
+        (*dataptr).use_routes = FALSE;
+        _kf_clear_key(kf, group, "ignore-auto-routes");
+        _kf_clear_key(kf, group, "never-default");
+    }
+    handle_generic_uint(kf, group, "route-metric", &(*dataptr).metric, NETPLAN_METRIC_UNSPEC);
+}
+
+static void
+parse_search_domains(GKeyFile* kf, const gchar* group, GArray** domains_arr)
+{
+    g_assert(domains_arr);
+    gsize len = 0;
+    gchar **split = g_key_file_get_string_list(kf, group, "dns-search", &len, NULL);
+    if (split) {
+        if (len == 0) {
+            _kf_clear_key(kf, group, "dns-search");
+            return;
```

Comment:
Ok, noob question: when uh, split is `!= NULL` but `len == 0`, does that mean something has been dynamically allocated? Do we still have to do `g_strfreev(split)` or there's nothing to be deallocated? This is a very unlikely case though I assume.

Reply:
From my understanding if split `!= NULL` and `len == 0` it means we're getting a pointer, pointing to `NULL` (i.e. the last element of an – empty – NULL-terminated list). So all we have is a pointer that will go out of scope at the end of the function, no dynamically allocated data that we need to free.

Well... we could put a `g_strfreev(split)` in there, just in case, but IMO it's not needed.

Still we want to make sure to clear the "dns-search=" line (which NM generates by default) from the keyfile, so it doesn't end up in the passthrough.